File: 2021/LevelUp_June/l002problems.cpp
#include <iostream>
#include <vector>

using namespace std;

int permutation(string str, string ans)
{
    if (str.length() == 0)
    {
        cout << (ans) << endl;
        return 1;
    }

    int count = 0;
    for (int i = 0; i < str.length(); i++)
    {
        char ch = str[i];
        string ros = str.substr(0, i) + str.substr(i + 1);
        count += permutation(ros, ans + ch);
    }

    return count;
}

int permutationUnique(string str, string ans)
{
    if (str.length() == 0)
    {
        cout << ans << endl;
        return 1;
    }

    vector<bool> vis(26, false);
    int count = 0;
    for (int i = 0; i < str.length(); i++)
    {
        char ch = str[i];
        if (!vis[ch - 'a'])
        {
            vis[ch - 'a'] = true;
            string ros = str.substr(0, i) + str.substr(i + 1);
            count += permutationUnique(ros, ans + ch);
        }
    }

    return count;
}

int permutationUnique2(string str, string ans)
{
    if (str.length() == 0)
    {
        cout << ans << endl;
        return 1;
    }

    char prev = '$';
    int count = 0;
    for (int i = 0; i < str.length(); i++)
    {
        char ch = str[i];
        if (prev != ch)
        {
            string ros = str.substr(0, i) + str.substr(i + 1);
            count += permutationUnique2(ros, ans + ch);
        }
        prev = ch;
    }

    return count;
}

int main()
{
    cout << permutationUnique2("aabb", "") << endl;
    return 0;
}

File: 2021/LevelUp_June/l002problems.java
public class l002problems {

    public static int equalSet(int[] arr, int idx, int sum1, int sum2, String set1, String set2) {
        if (idx == arr.length) {
            if (sum1 == sum2) {
                System.out.println(set1 + " = " + set2);
                return 1;
            }
            return 0;
        }

        int count = 0;
        count += equalSet(arr, idx + 1, sum1 + arr[idx], sum2, set1 + arr[idx] + " ", set2);
        count += equalSet(arr, idx + 1, sum1, sum2 + arr[idx], set1, set2 + arr[idx] + " ");

        return count;
    }

    public static void equalSet() {
        int[] arr = { 10, 20, 30, 40, 50, 60, 70, 80 };
        System.out.println(equalSet(arr, 1, arr[0], 0, arr[0] + " ", ""));
    }

    public static int permutation(String str, String ans) {
        if (str.length() == 0) {
            System.out.println(ans);
            return 1;
        }

        int count = 0;
        for (int i = 0; i < str.length(); i++) {
            char ch = str.charAt(i);
            String ros = str.substring(0, i) + str.substring(i + 1);
            count += permutation(ros, ans + ch);
        }

        return count;
    }

    public static void main(String[] args) {
        // equalSet();
        System.out.println(permutation("aaa", ""));
    }

}